src/core/qgsvectorlayerutils.cpp

```diff
@@ -908,55 +910,118 @@ QgsAttributeMap QgsVectorLayerUtils::QgsFeatureData::attributes() const
-      continue;
-    for ( QgsSymbolLayer *sl : item.originalSymbol()->symbolLayers() )
-    {
-      for ( const auto &r : sl->masks() )
+    public:
+      bool visitEnter( const QgsStyleEntityVisitorInterface::Node &node ) override
       {
-        refs[r.layerId()].insert( r.symbolLayerId() );
+        return ( node.type == QgsStyleEntityVisitorInterface::NodeType::SymbolRule );
       }
-    }
-  }
-  return refs;
+
+      bool visit( const QgsStyleEntityVisitorInterface::StyleLeaf &leaf ) override
+      {
+        if ( leaf.entity && leaf.entity->type() == QgsStyle::SymbolEntity )
+        {
+          auto symbolEntity = static_cast<const QgsStyleSymbolEntity *>( leaf.entity );
+          if ( symbolEntity->symbol() )
+          {
+            for ( QgsSymbolLayer *sl : const_cast<QgsSymbol *>( symbolEntity->symbol() )->symbolLayers() )
+            {
+              for ( const auto &r : sl->masks() )
+              {
+                masks[r.layerId()].insert( r.symbolLayerId() );
+              }
+            }
+          }
+        }
+        return true;
+      }
+
+      QHash<QString, QSet<QgsSymbolLayerId>> masks;
+  };
+
+  if ( ! layer->renderer() )
+    return {};
+
+  SymbolLayerMasksVisitor visitor;
+  layer->renderer()->accept( &visitor );
+  return std::move( visitor.masks );
 }
 
 void QgsVectorLayerUtils::setSymbolLayersEnabled( QgsFeatureRenderer *renderer, const QSet<QgsSymbolLayerId> &symbolLayerIds, bool enabled )
 {
-  for ( QgsLegendSymbolItem &item : renderer->legendSymbolItems() )
+  class EnableSymbolLayerVisitor : public QgsStyleEntityVisitorInterface
   {
-    int idx = 0;
-    if ( ! item.originalSymbol() )
-      continue;
-    for ( QgsSymbolLayer *sl : item.originalSymbol()->symbolLayers() )
-    {
-      if ( symbolLayerIds.contains( { item.ruleKey(), idx } ) )
+    public:
+      EnableSymbolLayerVisitor( const QSet<QgsSymbolLayerId> &layerIds, bool enable )
+        : mLayerIds( layerIds ), mEnable( enable )
+      {}
+
+      bool visitEnter( const QgsStyleEntityVisitorInterface::Node &node ) override
       {
-        sl->setEnabled( enabled );
+        if ( node.type == QgsStyleEntityVisitorInterface::NodeType::SymbolRule )
+        {
+          mCurrentRuleKey = node.identifier;
+          return true;
+        }
+        return false;
       }
-      idx++;
-    }
-  }
+
+      bool visit( const QgsStyleEntityVisitorInterface::StyleLeaf &leaf ) override
+      {
+        if ( leaf.entity && leaf.entity->type() == QgsStyle::SymbolEntity )
+        {
+          auto symbolEntity = static_cast<const QgsStyleSymbolEntity *>( leaf.entity );
+          if ( symbolEntity->symbol() )
+          {
+            int idx = 0;
+            for ( QgsSymbolLayer *sl : const_cast<QgsSymbol *>( symbolEntity->symbol() )->symbolLayers() )
+            {
+              if ( mLayerIds.contains( { mCurrentRuleKey + leaf.identifier, idx } ) )
+              {
+                sl->setEnabled( mEnable );
```

Comment:
This one makes me nervous - when I made the visitor, I deliberated whether it should be a read-only or editable visitor, and made the choice that editing brings a whole range of complications. So I based the whole approach on a read-only visitor. I fairly certain this decision will cause issues with trying to use it as an editable visitor, like you've done here.

So what I'm wondering -- if the point of this is to avoid rendering symbol layers in certain circumstances, couldn't that instead be done inside the render context? I.e. the render context stores the list of symbol layers to render, and the QgsSymbol rendering checks the context when determining whether a particular layer should be rendered (and prepared for used attributes, etc).  Seems much cleaner and safer to me, because you aren't editing the symbols themselves at all then.

Reply:
I've followed this approach initially, without success, but I did not remember why. I'll try harder, I agree this would be cleaner.